File: go/e2e/resilience_test.go
//go:build e2e

package e2e

import (
	"context"
	"errors"
	"sync/atomic"
	"testing"
	"time"

	goystore "github.com/goy-co/goy-store/go"
)

func TestResilience_CircuitBreakerAndRetries(t *testing.T) {
	metrics := goystore.DefaultMetrics()
	cfg := goystore.ResilienceConfig{
		MaxRetries:                 3,
		BaseBackoffMS:              10,
		CircuitBreakerThreshold:    3,
		CircuitBreakerResetSeconds: 1,
		OperationTimeoutSeconds:    2,
	}

	exec := goystore.NewResilientExecutor(cfg, metrics, "test", "mock")

	// 1. Retry test
	var attempts int32
	ctx := context.Background()

	res, err := goystore.ExecuteWithResilience(exec, ctx, "flaky_op", func(c context.Context) (int, error) {
		curr := atomic.AddInt32(&attempts, 1)
		if curr < 3 {
			return 0, errors.New("transient error")
		}
		return 42, nil
	})

	if err != nil || res != 42 {
		t.Fatalf("expected success with 42, got %v, err=%v", res, err)
	}
	if atomic.LoadInt32(&attempts) != 3 {
		t.Fatalf("expected 3 attempts, got %d", atomic.LoadInt32(&attempts))
	}

	// 2. Circuit breaker
	cb := goystore.NewCircuitBreaker(3, 200*time.Millisecond)
	if err := cb.CanExecute(); err != nil {
		t.Fatalf("cb should be closed initially: %v", err)
	}

	cb.OnFailure()
	cb.OnFailure()
	cb.OnFailure()

	// Circuit should be open
	if err := cb.CanExecute(); err == nil {
		t.Fatalf("cb should be open after 3 failures")
	}

	// Wait for reset timeout
	time.Sleep(250 * time.Millisecond)

	// Half-open probe
	if err := cb.CanExecute(); err != nil {
		t.Fatalf("cb should allow probe after reset timeout: %v", err)
	}

	cb.OnSuccess()
	if err := cb.CanExecute(); err != nil {
		t.Fatalf("cb should be closed after success: %v", err)
	}
}

File: rust/tests/resilience_test.rs
use goy_store::config::ResilienceConfig;
use goy_store::metrics::StoreMetrics;
use goy_store::resilience::{CircuitBreaker, ResilientExecutor};
use prometheus::Registry;
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::time::Duration;

#[tokio::test]
async fn test_resilience_circuit_breaker_and_retries() {
    let registry = Registry::new();
    let metrics = Arc::new(StoreMetrics::new(&registry).unwrap());

    let config = ResilienceConfig {
        max_retries: 3,
        base_backoff_ms: 10,
        circuit_breaker_threshold: 3,
        circuit_breaker_reset_seconds: 1,
        operation_timeout_seconds: 2,
    };

    let executor = ResilientExecutor::new(config, metrics.clone(), "test", "mock");

    // 1. Retry test
    let attempts = Arc::new(AtomicUsize::new(0));
    let attempts_clone = attempts.clone();

    let res = executor
        .execute("flaky_op", || {
            let att = attempts_clone.clone();
            async move {
                let current = att.fetch_add(1, Ordering::SeqCst);
                if current < 2 {
                    Err(anyhow::anyhow!("transient error"))
                } else {
                    Ok(42)
                }
            }
        })
        .await;

    assert!(res.is_ok());
    assert_eq!(res.unwrap(), 42);
    assert_eq!(attempts.load(Ordering::SeqCst), 3);

    // 2. Circuit breaker opening after threshold failures
    let cb = CircuitBreaker::new(3, Duration::from_millis(200));
    assert!(cb.can_execute().await.is_ok());

    cb.on_failure().await;
    cb.on_failure().await;
    cb.on_failure().await;

    // Circuit should be open
    assert!(cb.can_execute().await.is_err());

    // Wait for reset timeout
    tokio::time::sleep(Duration::from_millis(250)).await;

    // Probe should be allowed (half-open)
    assert!(cb.can_execute().await.is_ok());

    // On success, circuit closes
    cb.on_success().await;
    assert!(cb.can_execute().await.is_ok());
}
